### PR TITLE
Path module & JS tweaks

### PR DESCRIPTION
**Pull request by icidasset**
- Adds a `Path` module and `Path` type to reflect the changes in the Typescript SDK.
- Small refactor of the JS code, objects as params
- Allows setting custom webnative "factory" in JS

```elm
Wnfs.writeUtf8 base
  { path = Path.file [ "hello.txt" ]
  , tag = tagToString Mutation
  }
```

### CODE REVIEW

File: src/Webnative.elm [hunk 1]
```diff
@@ -109,10 +110,23 @@ type alias AppPermissions =
...
+    ```elm
+    import Webnative.Path as Path
+
+    { private =
+        [ Path.forPermissions (Path.directory [ "Audio", "Mixtapes" ])
+        , Path.forPermissions (Path.directory [ "Audio", "Playlists" ])
+        ]
+    , public =
+        []
+    }
```
Comment (matheus23) [inline src/Webnative.elm:123]: I see that using `Path t` is not an option here.

What do you think of
```elm
{ private =
  { files = [ ... ]
  , directories = [ ... ]
  }
, ...
}
```
?
Or maybe `privateFiles` and `privateDirectories`.

I don't think there's that much we gain by being as similar as possible to the typescript version. In the end I think ideally they diverge anyway, as the ts version tries to be nicely usable for typescript and the elm version tries to fit as good as possible into elm.

File: src/Webnative/Path.elm [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+    "{\"directory\":[\"foo\"]}"
+
+    >>> [ "bar" ]
+    ..>   |> file
+    ..>   |> toTypescriptFormat
+    ..>   |> Json.Encode.encode 0
+    "{\"file\":[\"bar\"]}"
+
+-}
+toTypescriptFormat : Path t -> Json.Value
```
Comment (matheus23) [inline src/Webnative/Path.elm:256]: I find reading typescript in the name of that function confusing. It doesn't seem to be related to typescript. I'd just go for a simple `Path.encode` name :)

File: src/Webnative/Path/Encapsulated.elm [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+toDirectory path =
+    case Path.kind path of
+        Directory ->
+            path
+                |> Path.unwrap
+                |> directory
+                |> Ok
+
+        File ->
+            Err "Encapsulated path is a File path, expected Directory"
```
Comment (matheus23) [inline src/Webnative/Path/Encapsulated.elm:31]: I don't think the error message adds much here.
The contrary, the Result type might indicate that there's multiple different errors that could happen.
Also, I worry that the strings will just get surfaced to a UI at some point. Instead, apps should provide their own, custom error message.
What if we just went `Path Encapsulated -> Maybe (Path Directory)`?

**Review by matheus23 — COMMENTED** [review]
I'm personally not a big fan of the `Path.Encapsulated` module, to be honest :S

I don't understand why it's needed. I'm sure you have reasons, so I'd love to discuss them.

I'd also suggest getting rid of `fromPosix`, as I personally really believe applications don't need the function, as they almost always expect either a directory or a file at some point.

---

I've been thinking a little bit about alternatives. What about this:
```elm
directoryFromPosix : String -> Result (Path Directory) (Path File)
fileFromPosix : String -> Result (Path File) (Path Directory)
```

Why the result? Initially I thought of only using e.g. `Maybe (Path File)`, but in the (I think) rare case you actually didn't care about whether it's a file or a directory, you'd have to match a third impossible case. So instead, we expose that they're absolutely one or the other in the types.

**Comment by matheus23** [comment]
Also: Man I love the :ghost: in the docs!

File: src/Webnative.elm (hunk 1 above)
Comment (icidasset) [reply src/Webnative.elm:123]: Oh ya, `private.files` is much better, I'll do that 👍

File: src/Webnative/Path/Encapsulated.elm (hunk 3 above)
Comment (icidasset) [reply src/Webnative/Path/Encapsulated.elm:31]: Sounds good 👍

**Comment by icidasset** [comment]
The reason I did this for `fromPosix` is that I want the users to be able to move around the result before deciding it needs to be a file or directory. For example, if I take a POSIX path as user input, I could add the parsed `Path Encapsulated` to my model. Whereas if I had to decide immediately if that's a directory or file, I wouldn't be able to do that.

I think we should keep the `Encapsulated` idea, just to have a generic way of moving around multiple kinds of `Path`s (something that isn't JSON). Also the handy thing with `Path Encapsulated` is that you could still use all `Path` functions on it.

We could of course have both, and add the `directoryFromPosix` and `fileFromPosix` functions in addition to `fromPosix`. If you think that would be useful I'll add them 👍 My plan was to publish this more generic implementation first, and then based on the usage see where we can add more specific functions and helper functions.

File: src/Webnative.elm [hunk 4]
```diff
@@ -115,18 +115,31 @@ type alias AppPermissions =
     import Webnative.Path as Path
 
     { private =
-        [ Path.forPermissions (Path.directory [ "Audio", "Mixtapes" ])
-        , Path.forPermissions (Path.directory [ "Audio", "Playlists" ])
-        ]
+        { directories = [ Path.directory [ "Audio", "Mixtapes" ] ]
+        , files = [ Path.directory [ "Audio", "Playlists", "Jazz" ] ]
```
Comment (matheus23) [inline src/Webnative.elm:119]: This should be `Path.file`, right? And maybe something like `Jazz.json`?

**Review by matheus23 — APPROVED** [review]
> I think we should keep the Encapsulated idea

> My plan was to publish this more generic implementation first, and then based on the usage see where we can add more specific functions and helper functions.

Yeah, my point is: I think it's easier to *add* things than to remove them.

> The reason I did this for fromPosix is that I want the users to be able to move around the result before deciding it needs to be a file or directory. For example, if I take a POSIX path as user input, I could add the parsed Path Encapsulated to my model. Whereas if I had to decide immediately if that's a directory or file, I wouldn't be able to do that.

If you need to do that, you could store a `String`, not a `Path Encapsulated`, right?

I get that storing a `String` doesn't sound nice, but I just want to make sure we're not introducing something that's going to be hard to remove later.

> We could of course have both, and add the directoryFromPosix and fileFromPosix functions in addition to fromPosix. If you think that would be useful I'll add them :+1: 

I would only introduce them as alternatives to having a `Path.Encapsulated` module. If we keep that module, we don't need them (so we only have one way of doing turning strings into dirs/files).

---

Anyway, I know I'm kind of pressing really hard here, so let me chill up a bit here :smile:

I'd be down to try this with `Path Encapsulated` in there, and get some dogfooding going from drive & other apps, while being open to any changes :+1:

File: src/Webnative.elm (hunk 4 above)
Comment (icidasset) [reply src/Webnative.elm:119]: Yup!

**Comment by icidasset** [comment]
> If you need to do that, you could store a String, not a Path Encapsulated, right?

Sure, but the cool thing with `Path Encapsulated` is that you can still use the `Path` module functions on it. So you can do `Path.kind encapsulatedPath` just as you would do `Path.kind filePath`.

> get some dogfooding going from drive & other apps, while being open to any changes

Yes exactly 👍
